oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2TokenAttributes.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.core;
+
+import org.springframework.util.Assert;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ *
+ * @author Clement Ng
+ *
+ */
+public class OAuth2TokenAttributes {
+	private Map<String, Object> attributes;
+
+	public OAuth2TokenAttributes() {
+		this.attributes = new HashMap<>();
+	}
+
+	public OAuth2TokenAttributes(Map<String, Object> attributes) {
+		Assert.notEmpty(attributes, "attributes cannot be empty");
+		this.attributes = attributes;
```

Comment:
Please make a copy of the attributes and make it an immutable copy.

Reply:
If we make the `attributes` map immutable via `Collections.unmodifiableMap`, I believe the `setUp()` method in the `OAuth2IntrospectionAuthenticationTokenTests` will have to be collapsed into the creation of the instance variables (since right now it puts values into an OAuth2.0 attributes map, but now we need the map to be prepopulated w/ data before constructing the `OAuth2.0TokenAttributes`). Would this be OK?

Reply:
Yes, that is fine. We do want to be careful when modifying tests, but it is better than breaking encapsulation.